Gpdma Channel Custom Configuration and Some Fixes

etiennecollin opened the pull request:
See https://github.com/embassy-rs/embassy/pull/3923 for details.

I am currently testing these changes. I don't think there are issues. Let me know what you think!

etiennecollin committed:
fix: modified dma channel state management

See
https://github.com/embassy-rs/embassy/pull/3923#discussion_r2094570176

etiennecollin committed:
fix: writing reserved bits

etiennecollin committed:
feat: custom DMA channel configuration

See
https://github.com/embassy-rs/embassy/pull/3923#issuecomment-2888810087

The default configuration of the channel which was done in `start()` is
now done in `new()` this allows overriding some settings through the new
`get_dma_channel` function. Only ringbuffers support this;
`LinkedListTransfer` and `Transfer` do not support that yet.

elagil commented:
Thanks, that makes a lot of sense! I will check against the RM and then we can merge it. Only the config/start split should be discussed further: https://github.com/embassy-rs/embassy/pull/3923#issuecomment-2889193736

We should probably find out if these changes also apply to the regular GPDMA transfer module that already exists, because I stole some configuration code from there.

etiennecollin committed:
feat: use provided TransferOptions instead of defaults

etiennecollin committed:
fix: suspend before reset

This follows the procedure outlined in the STM32U5 reference manual at
page 696.

etiennecollin commented:
Hi!
I answered in https://github.com/embassy-rs/embassy/pull/3923#issuecomment-2889283939. In the future, we can discuss the implementation here to avoid duplicating messages.

> We should probably find out if these changes also apply to the regular GPDMA transfer module that already exists, because I stole some configuration code from there.

I believe the extra configuration I want to introduce via `RegisterUpdaters` could be added to the regular GPDMA transfers, but since this PR is specifically for the linked-list + ringbuffer, I think we could work in another PR?

etiennecollin committed:
feat: custom dma configuration using RegisterUpdaters struct

See this PR comment:
https://github.com/embassy-rs/embassy/pull/3923#issuecomment-2889283939

etiennecollin committed:
fix: removed functions exposing channel registers

These functions could be used to cause UB.

etiennecollin committed:
fix: moved channel configuration from new() to start()

See this PR comment explaining why configuration in `new()` is a bad
idea:
https://github.com/embassy-rs/embassy/pull/3923#issuecomment-2889193736

etiennecollin commented:
Hello again,
Quick message to let you know that I tested my latest changes and everything seems to work properly.
Have a nice evening

elagil commented:
Thanks for the update! I will try to review your contribution this evening.

## Fixed GPDMA settings
I still don't quite get the use case of your register updaters. My original understanding was that you want to reconfigure the GPDMA at runtime, not only once in the beginning. Is your goal low-level control over GPDMA settings that are not yet covered by the simple `new_read`/`new_write` methods?

In that case, getting a mutable reference to the table, modifying it in the user application, and then configuring/starting the GPDMA with it seems like a simpler approach to me - you might have proposed that already.

## Variable GPDMA settings
Another use case is changing the configuration of the GPDMA at runtime. Let's take your keyboard matrix as an example. You might want to change something like the GPDMA source address after reading a row of keys. So your linked-list table may contain many entries, each of which having new settings. Wouldn't that be useful?

## Conclusion?
A possible solution would be to have another ring-buffer `new` method, e.g. `new_with_table` where a user provides a table directly. The table is then stored within the ring buffer struct. Note that we would have to make the ring buffer generic over the table length, or increase maximum table length. Both options aren't great. We could also store the table outside of the ring buffer and pass only its reference. But I'm afraid of the user moving around in memory by accident in that case.

etiennecollin commented:
Hello,
Thanks for getting back to me.

# Fixed and Variable  GPDMA settings

> My original understanding was that you want to reconfigure the GPDMA at runtime, not only once in the beginning.

Reconfiguring at runtime could be useful. In my case (with matrix scanning), I instead use two DMA channels. One to write and one to read. I order the transactions with a timer. This avoids having to reconfigure the DMA at runtime. You can see my working implementation here: https://github.com/etiennecollin/wave-rs/blob/6024fb410df77ce3884fe9101385273795d9537d/src/keyboard/dma.rs#L54-L102. It's my first time using the GPDMA, so I might have misunderstood, but are you saying I could be using a single GPDMA channel to both read and write the GPIO pins by having alternating read and write linked-list items?

> Is your goal low-level control over GPDMA settings that are not yet covered by the simple `new_read`/`new_write` methods?

Yes, I would sometimes need to change the individual registers (TR1, TR2, etc) of the linked-list items. This is not directly exposed to the `LinearItem::new_read()` and `LinearItem::new_write()` function parameters, but it would be possible to modify the item registers after they are returned. This leads to your next point:

> In that case, getting a mutable reference to the table, modifying it in the user application, and then configuring/starting the GPDMA with it seems like a simpler approach to me

That is true. I thought we might want to avoid exposing the table to the user, as modifying it can completely break the linked list. If you are ok with exposing it through a mutable reference, then that should be fine as we can modify its items.

> So your linked-list table may contain many entries, each of which having new settings. Wouldn't that be useful?

Yes it would.

# Conclusion

> Note that we would have to make the ring buffer generic over the table length

Yes, I don't think it's too bad. We simply need to put the `new()` function in a new `impl` bloc with `L=2`. This requires minimal changes to the codebase:

```rust
/// Ringbuffer for writing data using GPDMA linked-list mode.
pub struct WritableRingBuffer<'a, W: Word, const L: usize> {
    channel: Peri<'a, AnyChannel>,
    ringbuf: WritableDmaRingBuffer<'a, W>,
    table: Table<L>,
    options: TransferOptions,
}

impl<'a, W: Word> WritableRingBuffer<'a, W, 2> {
    /// Create a new ring buffer.
    pub unsafe fn new(
        channel: Peri<'a, impl Channel>,
        request: Request,
        peri_addr: *mut W,
        buffer: &'a mut [W],
        options: TransferOptions,
    ) -> Self {
        let channel: Peri<'a, AnyChannel> = channel.into();

        // Buffer halves should be the same length.
        let half_len = buffer.len() / 2;
        assert_eq!(half_len * 2, buffer.len());

        let items = [
            LinearItem::new_write(request, &mut buffer[..half_len], peri_addr),
            LinearItem::new_write(request, &mut buffer[half_len..], peri_addr),
        ];
        let table = Table::new(items);

        Self {
            channel,
            ringbuf: WritableDmaRingBuffer::new(buffer),
            table,
            options,
        }
    }
}

impl<'a, W: Word, const L: usize> WritableRingBuffer<'a, W, L> {
    /// Create a new ring buffer with a specific table.
    pub unsafe fn new_with_table(
        channel: Peri<'a, impl Channel>,
        buffer: &'a mut [W],
        options: TransferOptions,
        table: Table<L>,
    ) -> Self {
        let channel: Peri<'a, AnyChannel> = channel.into();

        Self {
            channel,
            ringbuf: WritableDmaRingBuffer::new(buffer),
            table,
            options,
        }
    }

	// ...
}
```

> We could also store the table outside of the ring buffer and pass only its reference. But I'm afraid of the user moving around in memory by accident in that case.

We could do this and it indeed is quite unsafe as, like you said, if the user moves the table, it's game over. I think it would be fair, since playing with the DMA is already quite _low-level_, to put the `start()` function as `unsafe` and add a big warning in the `rustdoc`.

This solution would also require the user wanting a simple ring-buffer to manually setup their table. If we opt for this, we could provide a `generate_simple_table()` function to generate the table that was originally generated in the `new()` function.

---

I think the two solutions you propose would work well. I would prefer the first alternative with a ring-buffer generic over the table size. To me, this is cleaner for usage, reduces the risks of panics/UB and still does not add too much bloat to the library. The mutable reference is _leaner_/cleaner to implement, but is not nearly as nice to use.

Let me know what you think.

> PS: If you are willing and it's convenient for you, we can hop on a quick voice call (Slack, Discord, Signal, Whatsapp, etc.) to discuss how we want to do thing. I'm available later tonight. Let me know if that's something you would want.

etiennecollin committed:
feat: add new_with_table() initializer for ring-buffers and removal of RegisterUpdaters

- It is now possible to pass a linked-list table to the ring-buffer with
the `new_with_table()` function or use the `new()` function for a basic
ring-buffer setup.
- A `simple_ring_buffer_table()` function was added to the read and
write ring-buffers to generate the same table as the one created by
`new()` in case the user only wants to customize the default table
options.
- RegisterUpdaters have been removed as the user now has direct access
to the table and its items if needed.

See: https://github.com/elagil/embassy/pull/1#issuecomment-2891997294

etiennecollin committed:
feat: use register wrappers instead of u32 for LinearItem

Since the register structs are no-field structs with
`repr(transparent)`, we can use them in the LinearItem with `repr(C)`.
This allows the user to call the convenient named setter functions for
the registers instead of manually changing the bits of the u32.

etiennecollin commented:
I implemented the first alternative with a ring-buffer generic over the table size to give you an idea of the changes needed. It works well, but let me know what you think. We can rollback the changes if needed or move to the second alternative if you prefer.

elagil commented:
> It's my first time using the GPDMA, so I might have misunderstood, but are you saying I could be using a single GPDMA channel to both read and write the GPIO pins by having alternating read and write linked-list items?

Yes, that is possible. The GPDMA will just reconfigure itself to whatever settings after completing one linked-list item. It doesn't require any use of the CPU, not even interrupt handling. You can run the GPDMA in certain CPU stop states as well (or at least the LPDMA).

> If you are ok with exposing it through a mutable reference, then that should be fine as we can modify its items.

I think I would be fine with it, because a user who sets up manual linked-list items probably knows enough details about the peripheral.

> Yes, I don't think it's too bad.

Implementing a generic linked-list length is not a problem, but it will increase flash usage when multiple different lengths are used in the program. I don't know how much overhead there will be after the optimizer does its thing, though.

> This solution would also require the user wanting a simple ring-buffer to manually setup their table.

This will be the main use of the GPDMA within embassy itself, as many drivers need a simple ring buffered DMA, usually implemented by means of the regular circular DMA.

By the way: At the moment, the ring buffers `impl`s across different DMA peripherals (e.g. regular DMA and GPDMA), for example the `new()` method, are implemented with the same interfaces. They don't implement a trait, so this is not obvious, but the build will break if the interfaces don't match.

> PS: If you are willing and it's convenient for you, we can hop on a quick voice call (Slack, Discord, Signal, Whatsapp, etc.) to discuss how we want to do thing. I'm available later tonight. Let me know if that's something you would want.

That sounds good!

etiennecollin committed:
fix: removed unnecessary mut reference

elagil commented on embassy-stm32/src/dma/gpdma/ringbuffered.rs:
```diff
@@ -63,10 +64,44 @@ impl<'a, W: Word> ReadableRingBuffer<'a, W> {
+        }
+    }
+
+    /// Create a new simple linked-list table.
+    ///
+    /// This uses two linked-list items, one for each half of the buffer.
+    pub unsafe fn simple_ring_buffer_table(request: Request, peri_addr: *mut W, buffer: &mut [W]) -> Table<2> {
```
As per STs nomenclature, we could call it

```suggestion
    /// Create a new simple linked-list table.
    ///
    /// This uses two linked-list items, one for each half of the buffer.
    pub unsafe fn new_ping_pong_table(request: Request, peri_addr: *mut W, buffer: &mut [W]) -> Table<2> {
```

etiennecollin replied:
Fixed in [92da825](https://github.com/elagil/embassy/pull/1/commits/92da82547453f8a6cb294619503eea7e05b54ff8)

elagil commented on embassy-stm32/src/dma/gpdma/ringbuffered.rs:
```diff
@@ -194,47 +227,75 @@ impl<'a, W: Word> Drop for ReadableRingBuffer<'a, W> {
+    /// Create a new simple linked-list table.
+    ///
+    /// This uses two linked-list items, one for each half of the buffer.
+    pub unsafe fn simple_ring_buffer_table(request: Request, peri_addr: *mut W, buffer: &[W]) -> Table<2> {
```
see above

etiennecollin replied:
Fixed in [92da825](https://github.com/elagil/embassy/pull/1/commits/92da82547453f8a6cb294619503eea7e05b54ff8)

etiennecollin committed:
fix: renamed simple table as per ST nomenclature

Co-authored-by: elagil <elagil@takanome.de>
